zerver/lib/outgoing_webhook.py

```diff
@@ -8,17 +8,46 @@
 import logging
 from six.moves import urllib
 from functools import reduce
+from requests import Response
 
 from django.utils.translation import ugettext as _
 
-from zerver.models import Realm, get_realm_by_email_domain, get_user_profile_by_id, get_client
+from zerver.models import Realm, UserProfile, get_realm_by_email_domain, get_user_profile_by_id, get_client
 from zerver.lib.actions import check_send_message
 from zerver.lib.queue import queue_json_publish
 from zerver.lib.validator import check_dict, check_string
 from zerver.decorator import JsonableError
 
 MAX_REQUEST_RETRIES = 3
 
+class OutgoingWebhookServiceInterface(object):
+
+    def __init__(self, base_url, token, user_profile, service_name):
+        # type: (Text, Text, UserProfile, Text) -> None
+        self.base_url = base_url  # type: Text
+        self.token = token  # type: Text
+        self.user_profile = user_profile  # type: Text
+        self.service_name = service_name  # type: Text
+
+    # would receive the data generated on triggering of outgoing webhooks like service name, trigger method, message
+    # object, etc and return rest operation related info like relative url, base url, method of query i.e. POST, GET,
+    # etc and data to be sent to third party url.
+    def process_command(self, event):
+        # type: (Dict[str, Any]) -> Tuple[Dict[str ,Any], Dict[str, Any]]
+        raise NotImplementedError()
+
+    # in case of success of request, it would parse the response obtained and return the content of message
+    # to be sent to user
+    def process_response(self, response, event):
+        # type: (Response, Dict[Text, Any]) -> Dict[str, Any]
+        raise NotImplementedError()
+
+    # in case of failure of request, it would parse the failure response obtained and return the content of message
+    # to be sent to user
+    def handle_failure(self, response, event):
```

Comment:
Style nit: `process` and `handle` are very generic verbs that can mean a lot of different things.  Would it make sense to rename these methods to something much more specific, such as `message_for_rest_success` and `message_for_rest_failure`?  (Or `message_for_rest_response` if you want to just have a single method?)